workspaces/libnpmexec/lib/with-lock.js

```diff
@@ -0,0 +1,162 @@
+const { onExit } = require('signal-exit')
+
+// a lockfile implementation inspired by the unmaintained proper-lockfile library
+//
+// similarities:
+// - based on mkdir's atomicity
+// - works across processes and even machines (via NFS)
+// - cleans up after itself
+// - detects compromised locks
+//
+// differences:
+// - higher-level API (just a withLock function)
+// - written in async/await style
+// - uses mtime + inode for more reliable compromised lock detection
+// - more ergonomic compromised lock handling (i.e. withLock will reject, and callbacks have access to an AbortSignal)
+// - uses a more recent version of signal-exit
+
+const touchInterval = 100
+// mtime precision is platform dependent, so use a reasonably large threshold
+const staleThreshold = 5_000
+
+// track current locks and their cleanup functions
+const currentLocks = new Map()
+
+function cleanupLocks () {
+  for (const [, cleanup] of currentLocks) {
+    try {
+      cleanup()
+    } catch (err) {
+      //
+    }
+  }
+}
+
+// clean up any locks that were not released normally
+onExit(cleanupLocks)
+
+/**
+ * Acquire an advisory lock for the given path and hold it for the duration of the callback.
+ *
+ * The lock will be released automatically when the callback resolves or rejects.
+ * Concurrent calls to withLock() for the same path will wait until the lock is released.
+ */
+async function withLock (filePath, cb) {
+  const lockPath = `${filePath}.lock`
+  try {
+    const signal = await acquireLock(lockPath)
+    return await new Promise((resolve, reject) => {
+      signal.addEventListener('abort', () => {
+        reject(Object.assign(new Error('Lock compromised'), { code: 'ECOMPROMISED' }))
+      });
+
+      (async () => {
+        try {
+          resolve(await cb(signal))
+        } catch (err) {
+          reject(err)
+        }
+      })()
+    })
+  } finally {
+    await releaseLock(lockPath)
+  }
+}
+
+function acquireLock (lockPath) {
+  const operation = retry.operation({
+    minTimeout: 100,
+    maxTimeout: 5_000,
+    // if another process legitimately holds the lock, wait for it to release;
+    // if it dies abnormally and the lock becomes stale, we'll acquire it automatically
+    forever: true,
+  })
+  return new Promise((resolve, reject) => {
+    operation.attempt(async () => {
+      try {
+        await fs.mkdir(lockPath)
+      } catch (err) {
+        if (err.code !== 'EEXIST') {
+          return reject(err)
+        }
+
+        try {
+          const status = await getLockStatus(lockPath)
+
+          if (status === 'locked') {
+            // let's see if we can acquire it on the next attempt 🤞
+            return operation.retry(err)
+          }
+          if (status === 'stale') {
+            // there is a very tiny window where another process could also release the stale lock and acquire it
+            // before we release it here; the lock compromise checker should detect this and throw an error
```

Comment:
```suggestion
            // there is a very tiny window where another process could also release the stale lock and acquire it before we release it here; the lock compromise checker should detect this and throw an error
```

We recently changed our linting rules to no longer complain about long comments.  Breaking them up like this is less accessible than letting them go on naturally like prose is supposed to.

As far as the contents of the comment: we are moving this race condition problem from a situation where it happens accidentally during reification with no way to control it, to a situation where it's much less likely to happen here, and in a way that lets us handle it MUCH more gracefully.  👍 on doing this.